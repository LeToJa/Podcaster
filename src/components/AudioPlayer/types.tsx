export interface AudioPlayerTypes {
    url: string
    type: string
}