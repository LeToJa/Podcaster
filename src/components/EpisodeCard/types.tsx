export interface EpisodeCardTypes {
    podcastId: string
    episodeId: number
    title: string
    date: string
    duration: string
}