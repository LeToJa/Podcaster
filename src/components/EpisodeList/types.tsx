import { RSSResponse } from '../../helpers/RSSParser/types'

export interface EpisodeListTypes {
    podcastId: string
    episodes: RSSResponse[] | false
}