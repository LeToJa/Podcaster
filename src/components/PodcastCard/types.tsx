export interface PocastCardTypes {
    id: string
    name: string
}