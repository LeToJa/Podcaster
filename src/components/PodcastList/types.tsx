import { PocastCardTypes } from '../PodcastCard/types'

export interface PodcastListTypes {
    podcasts: PocastCardTypes[]
}