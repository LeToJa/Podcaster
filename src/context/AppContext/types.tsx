export interface AppContextTypes {
    loading: boolean
    toggleLoading: (value: boolean) => void
}