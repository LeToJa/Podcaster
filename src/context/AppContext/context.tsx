import { createContext, useState } from 'react'

import { ReactElementWithChildrenTypes } from '../../types'
import { AppContextTypes } from './types'

export const AppContext = createContext<AppContextTypes | undefined>(undefined)

export const AppProvider = ({ children }: ReactElementWithChildrenTypes) => {
    const [loading, setLoading] = useState<boolean>(true)

    const toggleLoading = (value: boolean) => setLoading(value)

    return <>
        <AppContext.Provider value={{ loading, toggleLoading }}>
            {children}
        </AppContext.Provider>
    </>
}