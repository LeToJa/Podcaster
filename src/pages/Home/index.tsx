import { useEffect, useState } from 'react'

import { useApp } from '../../context/AppContext'

import { getPodcastList } from '../../helpers/ItunesAPI'

import PodcastCard from '../../components/PodcastCard'
import { PocastCardTypes } from '../../components/PodcastCard/types'

const Home = () => {
    const { loading, toggleLoading } = useApp()

    const [filter, setFilter] = useState<string>('')
    const [podcasts, setPodcasts] = useState<PocastCardTypes[] | undefined>([])

    useEffect(() => {
        const fetchPodcasts = async () => {
            const content = await getPodcastList()

            toggleLoading(false)
            setPodcasts(content)
        }

        toggleLoading(true)
        fetchPodcasts().catch(err => console.error(err))
    }, [])


    const filteredPodcasts = podcasts?.filter(podcasts =>
        podcasts.title.toLowerCase().includes(filter.toLowerCase()) ||
        podcasts.author.toLowerCase().includes(filter.toLowerCase())
    )

    if (!loading) return <>
        <form className='flex justify-center mb-6'>
            <input
                type='text'
                placeholder='Search podcasts'
                name='filter-podcasts'
                id='filter-podcasts'
                className='bg-gray-100 px-4 py-3 rounded shadow-md outline-sky-600'
                value={filter}
                onChange={(e) => setFilter(e.target.value)}
            />
        </form>
        <section className='grid grid-cols-5 gap-4'>
            {
                filteredPodcasts?.length ?
                    filteredPodcasts?.map((podcast, key) => <PodcastCard key={key} {...podcast} />) :
                    <div className='col-span-5 font-bold text-xl text-gray-700 text-center'>No podcasts :(</div>
            }
        </section>
    </>
}

export default Home