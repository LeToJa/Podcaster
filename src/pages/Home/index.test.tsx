import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import { describe, it, expect, vi, beforeEach } from 'vitest'
import { BrowserRouter } from 'react-router-dom'

import { AppContext } from '../../context/AppContext/context'
import { getPodcastList } from '../../helpers/ItunesAPI'
import { AppContextProps } from '../../context/AppContext/types'
import { ReactElementWithChildrenTypes } from '../../types'

import Home from '.'

vi.mock('../../helpers/ItunesAPI', () => ({
    getPodcastList: vi.fn(),
}))

const mockAppContext = {
    loading: false,
    toggleLoading: vi.fn()
}

const MockAppProvider = ({ loading, toggleLoading, children }: AppContextProps & ReactElementWithChildrenTypes) => {
    return (
        <AppContext.Provider value={{ loading, toggleLoading }}>
            {children}
        </AppContext.Provider>
    )
}

const mockPodcasts = [
    {
        id: '1',
        title: 'Podcast One',
        author: 'Author One',
        artwork: 'https://via.placeholder.com/150',
    },
    {
        id: '2',
        title: 'Podcast Two',
        author: 'Author Two',
        artwork: 'https://via.placeholder.com/150',
    },
]

const mockGetPodcastList = vi.mocked(getPodcastList)

describe('Home', () => {
    beforeEach(() => {
        mockGetPodcastList.mockResolvedValue(mockPodcasts)
    })

    it('renders correctly and matches snapshot', () => {
        const { container } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <Home />
                </MockAppProvider>
            </BrowserRouter>
        )

        expect(container).toMatchSnapshot()
    })

    it('displays podcasts after fetch', async () => {
        const { getByText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <Home />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText('Podcast One')).toBeTruthy()
            expect(getByText('Podcast Two')).toBeTruthy()
        })
    })

    it('filters podcasts based on user input', async () => {
        const { getByText, queryByText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <Home />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText('Podcast One')).toBeTruthy()
            expect(getByText('Podcast Two')).toBeTruthy()
        })

        fireEvent.change(screen.getByPlaceholderText('Search podcasts'), {
            target: { value: 'One' },
        })

        expect(getByText('Podcast One')).toBeTruthy()
        expect(queryByText('Podcast Two')).toBeFalsy()
    })

    it('displays "No podcasts :(" when no podcasts match the filter', async () => {
        const { getByText, getByPlaceholderText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <Home />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText('Podcast One')).toBeTruthy()
            expect(getByText('Podcast Two')).toBeTruthy()
        })

        fireEvent.change(getByPlaceholderText('Search podcasts'), {
            target: { value: 'Nonexistent Podcast' },
        })

        expect(getByText('No podcasts :(')).toBeTruthy()
    })

    it('displays "No podcasts :(" when the API returns no podcasts', async () => {
        mockGetPodcastList.mockResolvedValue([])

        const { getByText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <Home />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText('No podcasts :(')).toBeTruthy()
        })
    })
})
