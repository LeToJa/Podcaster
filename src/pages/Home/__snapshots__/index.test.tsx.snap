// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > renders correctly and matches snapshot 1`] = `
<div>
  <form
    class="flex justify-center mb-6"
  >
    <input
      class="bg-gray-100 px-4 py-3 rounded shadow-md outline-sky-600"
      id="filter-podcasts"
      name="filter-podcasts"
      placeholder="Search podcasts"
      type="text"
      value=""
    />
  </form>
  <section
    class="grid grid-cols-5 gap-4"
  >
    <section
      class="col-span-5 font-bold text-xl text-gray-700 text-center"
    >
      No podcasts :(
    </section>
  </section>
</div>
`;
